### PR TITLE
Optionally generate intelligent release logs using Glyph

### PR DESCRIPTION
**Pull request by tushar7sharma**
DO NOT MERGE UNTIL GLYPH IS RELEASED AND CAN BE IMPORTED BY KEBECHET

- Version Manager can now read 3 more parameters: 

`changelog_smart: false` : Controls if Intelligent Logs need to be generated
`changelog_classifier: fasttext` : Specifies which ML classifier to use (Defaults to fasttext, if not provided)
`changelog_format: cluster_similar` : Specifies which Formatter to use (Defaults to cluster_similar, if not provided)

- Glyph's `generate_log(
                changelog,
                Format.by_name(changelog_format),
                MLModel.by_name(changelog_classifier),)` method can be used for generating smart log.

- `from thoth.glyph import generate_log`
`from thoth.glyph import MLModel`
`from thoth.glyph import Format`
These methods/constants are imported from Glyph. Therefore it's added in the requirements.txt file

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -10,3 +10,4 @@ semver
 sentry-sdk
 thoth-common
 thoth-sourcemanagement
+thoth-glyph
```
Comment (saisankargochhayat) [inline requirements.txt:13]: This should also be added to the pipfile

**Comment by saisankargochhayat** [comment]
Please add these parameters in the version manager readme :smiley:

File: kebechet/managers/version/version.py [hunk 2]
```diff
@@ -255,9 +263,24 @@ def _compute_changelog(
...
-        ).splitlines()
+        _LOGGER.info("Smart Log : " + str(changelog_smart))
+
+        if changelog_smart:
+            _LOGGER.info("Classifier : " + changelog_classifier)
+            _LOGGER.info("Format : " + changelog_format)
+            changelog = repo.git.log(
+                f"{old_version}..HEAD", no_merges=True, format="%s"
+            ).splitlines()
+            changelog = generate_log(
```
Comment (fridex) [inline kebechet/managers/version/version.py:274]: If you want to be precise, you can catch any exception from the thoth-glyph library (the top level exception from the library's exception hierarchy) and report to user (via closing issue comment) that there is something wrong in the config (with the exception text).

File: requirements.txt (hunk 1 above)
Comment (tushar7sharma) [reply requirements.txt:13]: Fixed

File: kebechet/managers/version/version.py [hunk 3]
```diff
@@ -255,9 +263,24 @@ def _compute_changelog(
...
             old_version = repo.git.rev_list("HEAD", max_parents=0)
 
-        changelog = repo.git.log(
-            f"{old_version}..HEAD", no_merges=True, format="* %s"
-        ).splitlines()
+        _LOGGER.info("Smart Log : " + str(changelog_smart))
+
+        if changelog_smart:
+            _LOGGER.info("Classifier : " + changelog_classifier)
+            _LOGGER.info("Format : " + changelog_format)
```
[the remark this answers is not in the document]
Comment (tushar7sharma) [reply kebechet/managers/version/version.py]: Fixed

File: kebechet/managers/version/version.py (hunk 2 above)
Comment (tushar7sharma) [reply kebechet/managers/version/version.py:274]: Catching Thoth's exceptions now, and closing the issue with a comment (with the exception text).

**Comment by tushar7sharma** [comment]
> Please add these parameters in the version manager readme

Done

File: kebechet/managers/version/version.py [hunk 4]
```diff
@@ -226,7 +228,13 @@ def _is_release_request(issue_title):
 
     @staticmethod
     def _compute_changelog(
-        repo: Repo, old_version: str, new_version: str, version_file: bool = False
+        repo: Repo,
+        old_version: str,
+        new_version: str,
+        changelog_smart: bool,
+        changelog_classifier: str,
+        changelog_format: str,
```
Comment (fridex) [inline kebechet/managers/version/version.py:236]: Are these two optional?

```python
        changelog_classifier: Optional[str],
        changelog_format: Optional[str],
```

If so a check would be appropriate.

File: kebechet/managers/version/version.py [hunk 5]
```diff
@@ -386,10 +412,22 @@ def run(  # type: ignore
                 if not version_identifier:
                     _LOGGER.error("Giving up with automated release")
                     return
+
```
Comment (fridex) [inline kebechet/managers/version/version.py:415]: No need for another blank line.

File: kebechet/managers/version/version.py (hunk 4 above)
Comment (tushar7sharma) [reply kebechet/managers/version/version.py:236]: No they aren't optional.
The arguments will always be passed on from the run() function. And run() makes sure that default classifier and formatter are selected if caller fails to specify them.

File: kebechet/managers/version/version.py (hunk 5 above)
Comment (tushar7sharma) [reply kebechet/managers/version/version.py]: Fixed.

File: kebechet/managers/version/version.py (hunk 4 above)
Comment (fridex) [reply kebechet/managers/version/version.py:236]: > The arguments will always be passed on from the run() function. And run() makes sure that default classifier and formatter are selected if caller fails to specify them.

Where is this happening?

Comment (tushar7sharma) [reply kebechet/managers/version/version.py:236]: https://github.com/thoth-station/kebechet/blob/8a711e03e95d3026174eacf40585d63d028dc447/kebechet/managers/version/version.py#L339-L348

Comment (fridex) [reply kebechet/managers/version/version.py:236]: The input is passed from `.thoth.yaml` file sitting inside the repo. There is no check done on the actual keys present. One can easily produce wrong entries (e.g. `changelog_foo`) which will cause issues on kebechet side (and no response on the user side). I think we can introduce a mechanism to tackle this.

File: kebechet/managers/version/version.py [hunk 6]
```diff
@@ -387,9 +413,20 @@ def run(  # type: ignore
...
+                        changelog_format,
+                        version_file=changelog_file,
+                    )
+                except ThothGlyphException as exc:
+                    _LOGGER.exception("Failed to generate smart release log")
+                    issue.comment(str(exc))
+                    issue.close()
 
                 # If an issue exists, we close it as there is no change to source code.
                 if not changelog:
```
Comment (fridex) [inline kebechet/managers/version/version.py:432]: Will be the comment below put as well? I think there is a missing `return` statement on error.

**Comment by fridex** [comment]
@tushar7sharma please rebase

**Comment by saisankargochhayat** [comment]
Can we also ensure, pre-commit issues are solved or ignored?

File: kebechet/managers/version/version.py (hunk 6 above)
Comment (tushar7sharma) [reply kebechet/managers/version/version.py:434]: Fixed

**Comment by tushar7sharma** [comment]
> Can we also ensure, pre-commit issues are solved or ignored?

I think the pre-commit issue is raised because Glyph isn't installed and it's functions can't be imported

**Comment by fridex** [comment]
> > Can we also ensure, pre-commit issues are solved or ignored?
> 
> I think the pre-commit issue is raised because Glyph isn't installed and its functions can't be imported

It looks like it uses version 0.0.0 which I created just as a placeholder on PyPI. I've triggered a new release in https://github.com/thoth-station/glyph/issues/32

**Comment by tushar7sharma** [comment]
@fridex Merged with upstream/master to fix conflicts.

**Comment by goern** [comment]
ping_!

**Comment by fridex** [comment]
> ping_!

Let's merge this and I'll create an update in a PR.

**Pull request merged**